### ISSUE TITLE
Training of baseline models for 2m temperature downscaling task

### ISSUE THREAD

**Issue by mlangguth89** [mlangguth89/downscaling_benchmark#48]
In this issue branch, training of the baseline model for the T2m downscaling task will be documented.
The related dataset is available here:

- high-res input data (ERA5 data bilinearly upsampled on target grid): `/p/scratch/hclimrep/maelstrom/downscaling/downscaling_benchmark_dataset/benchmark_t2m/dataset/hres_input`
- low-res input data (ERA5 on coarse grid): `/p/scratch/hclimrep/maelstrom/downscaling/downscaling_benchmark_dataset/benchmark_t2m/dataset/coarse_input`
The corresponding configuration files for the datasets are available here:
- high-res input data: `<base_dir_repo>/config/datasets/config_ds_t2m_hres_input.json`
- low-res input data: `<base_dir_repo>/config/datasets/config_ds_t2m_coarse_input.json` 

The following models should be trained three times from which the best performing model in terms of RMSE and spatial variability (similar to #23) should be selected:
- [x] Sha U-Net
- [x] Sha WGAN
- [x] DeepRU
- [x] SwinIR
- [x] Harris WGAN

The respective configuration files are placed under `<base_dir_repo>/config/models`.

⚠ **Please ensure the following** ⚠ 
 - Share your results in comments of this issue-branch!
 - Copy your benchmark model checkpoints to this directory `/p/project/hclimrep/downscale_bench/trained_models/t2m`
- Copy your results (= output of some basic postprocessing with RMSE, Bias, gradient amplitude reatio etc.) to this directory `/p/project/hclimrep/downscale_bench/trained_results/t2m`

**Comment by seblehner** [comment]
### Experiment 1: Harris WGAN, RW2000

| Epoch | duration [t] | c_loss | gp_loss | d_loss | cg_loss | recon_loss | g_loss | val_cg_loss | val_recon_loss |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 4012/4012 | 4086s |  8.5381 |  8.4521 |  16.9903 |  -7.6184e+01 |  71.4426 |  -4.7414e+00 |  -1.5830e+02 |  108.5209 |
| 4012/4012 | 4013s |  -2.3580e+00 |  8.9723 |  6.6143 |  -1.0072e+02 |  63.6386 |  -3.7080e+01 |  -1.7143e+01 |  62.3816 |
| 4012/4012 | 4015s |  -8.0364e+00 |  4.8921 |  -3.1442e+00 |  -5.8968e+01 |  65.0362 |  6.0679 |  -6.3663e+01 |  46.5348 |
| 4012/4012 | 4006s |  3.7983 |  8.4273 |  12.2256 |  -2.5450e+02 |  48.7144 |  -2.0579e+02 |  -2.5887e+02 |  52.2391 |
| 4012/4012 | 4009s |  0.5659 |  9.2382 |  9.8041 |  -8.8029e+01 |  52.1969 |  -3.5832e+01 |  -8.6894e+01 |  42.8119 |
| 4012/4012 | 4010s |  -3.3264e-01 |  7.8941 |  7.5614 |  -2.3169e+02 |  56.6884 |  -1.7500e+02 |  -2.3906e+02 |  39.4633 |
| 4012/4012 | 4010s |  1.7925 |  8.5228 |  10.3153 |  -1.6633e+02 |  48.9212 |  -1.1741e+02 |  -1.7240e+02 |  47.9663 |


### Experiment 2:  Harris WGAN, RW500

| Epoch | duration [t] | c_loss | gp_loss | d_loss | cg_loss | recon_loss | g_loss | val_cg_loss | val_recon_loss |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 4012/4012 | 4062s |  -9.8792e-01 |  8.1599 |  7.1720 |  -4.6814e+02 |  36.0037 |  -4.3214e+02 |  -4.8221e+02 |  24.1294 |
| 4012/4012 | 3984s |  0.4163 |  9.0613 |  9.4776 |  -2.1059e+03 |  21.7100 |  -2.0842e+03 |  -2.1947e+03 |  26.1731 |
| 4012/4012 | 3985s |  -3.2715e-01 |  8.9957 |  8.6686 |  -2.9600e+03 |  13.0327 |  -2.9470e+03 |  -3.0472e+03 |  17.3651 |
| 4012/4012 | 3986s |  -6.3770e-01 |  9.0733 |  8.4356 |  -3.8508e+03 |  12.2251 |  -3.8386e+03 |  -3.8151e+03 |  13.9362 |
| 4012/4012 | 3985s |  0.0620 |  9.1971 |  9.2591 |  -4.1687e+03 |  11.1875 |  -4.1575e+03 |  -4.0902e+03 |  10.4300 |
| 4012/4012 | 3983s |  0.0532 |  9.1610 |  9.2142 |  -4.4424e+03 |  10.7013 |  -4.4317e+03 |  -4.4145e+03 |  13.2123 |
| 4012/4012 | 3982s |  0.1050 |  9.1164 |  9.2214 |  -4.7486e+03 |  11.5158 |  -4.7370e+03 |  -4.9056e+03 |  14.4554 |

**Comment by mlangguth89** [comment]
In previous experiments without batch normalization and a missing residual block in the critic, but with `channels_start=512` (instead of 128 which is currently used due to memory restrictions), the following loss terms have been reported during training:
| Epoch | duration [t] | c_loss | gp_loss | d_loss | cg_loss | recon_loss | g_loss | val_cg_loss | val_recon_loss |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
|4012/4012 | 7480s | 4.8610e-01 | 3.0955 | 2.6094  | 46.6046 | 16.8405 | 63.4451 | 55.7711 | 14.9473 |

These results in RMSE-values during evaluation of about 1.2 K which is about 10% smaller than with the results reported above.

Currently, efforts are on-going to allow for a larger number of filters in the critic with mixed precision. The number of filters in the critics is argued to be crucial in the original Harris et al., 2022-paper.

**Comment by mlangguth89** [comment]
Mixed precision seems to be a bit critic. Using losses at tf.float16-precision (incl. scaling and unscaling as documented [here](https://keras.io/2.16/api/mixed_precision/loss_scale_optimizer/)) resulted into exploding gradients, even when manually choosing smaller initial scale values. 
When doing all gradient calculations at tf.float32-precision, thraining dynamics could be largely stabilized, but still become unstable (aka exploding gradients) as documented below in anexperiment with `channels_start=256`:
After the first epoch, the following losses have been present:
```4012/4012 - 3662s - c_loss: -1.2500e+00 - gp_loss: 8.8240 - d_loss: 7.5740 - cg_loss: -5.7602e+03 - recon_loss: 40.7719 - g_loss: -5.7195e+03 - val_cg_loss: -5.7960e+03 - val_recon_loss: 51.2561```
However, during the 2nd epoch, gradients exploded, yielding:
```
4012/4012 - 3561s - c_loss: nan - gp_loss: 6.9925 - d_loss: nan - cg_loss: -inf - recon_loss: 950.7205 - g_loss: -inf - val_cg_loss: -inf - val_recon_loss: 3102.2668
```

**Comment by seblehner** [comment]
### Experiment 3: Harris WGAN, recon_weight=2000, filters critic 256, batch_size=8

Results seem to be comparable to the experiment 1 above with `recon_weight=2000`, but not better than `recon_weight=500`, so the increasing number of filters in the critic either had little effect here, or any positive effect is dampened by the reduced `batch_size=8` (instead of `batch_size=16`). The reduction in `batch_size` is necessary because the model is too big to fit into memory otherwise.

| Epoch | duration [t] | c_loss | gp_loss | d_loss | cg_loss | recon_loss | g_loss | val_cg_loss | val_recon_loss |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 4012/4012 | 2933s |  2.4840 |  8.1544 |  10.6384 |  -1.0005e+02 |  82.2260 |  -1.7827e+01 |  -1.4178e+02 |  111.5103 |
| 4012/4012 | 2850s |  -1.7180e+00 |  8.4077 |  6.6897 |  167.6420 |  99.9190 |  267.5609 |  116.4286 |  88.1120 |
| 4012/4012 | 2850s |  5.6754 |  5.2705 |  10.9458 |  -2.5739e+02 |  70.3684 |  -1.8702e+02 |  -2.9085e+02 |  71.8593 |
| 4012/4012 | 2853s |  1.1245 |  5.4214 |  6.5459 |  -5.0663e+02 |  49.6637 |  -4.5696e+02 |  -4.9158e+02 |  46.2619 |
| 4012/4012 | 2850s |  4.0907 |  5.0205 |  9.1111 |  -3.6688e+02 |  47.8677 |  -3.1901e+02 |  -3.7203e+02 |  54.5602 |
| 4012/4012 | 2851s |  1.3451 |  7.0107 |  8.3558 |  -1.6379e+02 |  53.7667 |  -1.1002e+02 |  -1.3695e+02 |  53.9035 |
| 4012/4012 | 2854s |  1.3118 |  8.0505 |  9.3623 |  -8.3397e+01 |  43.5894 |  -3.9807e+01 |  -8.7258e+01 |  44.2000 |
| 4012/4012 | 2851s |  -3.6837e+00 |  3.9973 |  0.3135 |  -1.9799e+02 |  45.3964 |  -1.5259e+02 |  -2.5008e+02 |  62.3615 |
| 4012/4012 | 2853s |  0.0387 |  2.2749 |  2.3136 |  -2.0138e+02 |  60.5588 |  -1.4082e+02 |  -1.2672e+02 |  47.0857 |
| 4012/4012 | 2850s |  -8.6792e-02 |  1.5039 |  1.4172 |  -1.5368e+02 |  40.8194 |  -1.1286e+02 |  -1.6185e+02 |  45.3543 |
| 4012/4012 | 2852s |  -1.4953e+00 |  2.0436 |  0.5483 |  -1.4033e+02 |  48.7716 |  -9.1555e+01 |  -2.1671e+02 |  60.3041 |
| 4012/4012 | 2851s |  -6.2867e+00 |  2.8198 |  -3.4669e+00 |  -6.3466e+01 |  47.7640 |  -1.5703e+01 |  -2.1698e+01 |  53.4653 |
| 4012/4012 | 2848s |  2.6066 |  2.3780 |  4.9846 |  20.0689 |  53.3793 |  73.4482 |  -1.7570e+00 |  43.9439 |
| 4012/4012 | 2851s |  5.0710 |  3.0988 |  8.1698 |  -7.8342e+02 |  49.6747 |  -7.3375e+02 |  -7.6847e+02 |  50.5957 |
| 4012/4012 | 2849s |  2.6373 |  5.4795 |  8.1168 |  -4.3257e+02 |  35.9648 |  -3.9661e+02 |  -4.5233e+02 |  41.1990 |
| 4012/4012 | 2849s |  0.3826 |  1.9704 |  2.3530 |  -4.4168e+02 |  38.8834 |  -4.0280e+02 |  -4.3488e+02 |  37.3820 |
| 4012/4012 | 2850s |  -1.3635e+00 |  4.8746 |  3.5111 |  -2.6135e+02 |  44.5904 |  -2.1676e+02 |  -2.5743e+02 |  42.4871 |
| 4012/4012 | 2851s |  -2.1385e+00 |  3.4208 |  1.2823 |  -1.4063e+02 |  45.0192 |  -9.5610e+01 |  -5.4885e+01 |  33.6285 |
| 4012/4012 | 2850s |  0.3713 |  0.7000 |  1.0713 |  -6.6886e+01 |  46.8565 |  -2.0029e+01 |  -8.8740e+01 |  38.1617 |
| 4012/4012 | 2851s |  -1.9677e+00 |  2.7361 |  0.7684 |  -6.7976e+01 |  35.7133 |  -3.2262e+01 |  -4.5454e+01 |  35.0685 |

**Comment by mlangguth89** [comment]
After numerous tests, we (Sebastian and me) came up with two 'skilful' Harris WGAN downscaling model versions, named `harris_wgan_benchmark4_t2m_rw1000_filter256
` and `harris_wgan_benchmark_cs256_4gpus_layernorm_higherlr`. These two model differ in their model configuration.
**_harris_wgan_benchmark4_t2m_rw1000_filter256_**: 

- `batch_size = 8`
- `norm='batch'` for generator

**_harris_wgan_benchmark_cs256_4gpus_layernorm_higherlr_**
- `batch_size = 32` (=4*8)
- `norm='layer'` for generator and critic (!)
Both models use 256 filters in the initial convolution layer of the critic (opposed to 512 in the original Harris WGAN model). 

In terms of the 'headline' scores RMSE and 'Gradient amplitude ratio', both model perform similar evaluated on the whole test dataset. For the ensemble mean, we get:
**_harris_wgan_benchmark4_t2m_rw1000_filter256_**: 
- `RMSE=1.2073 K`
- `grad_ampl_ratio=0.8545`

**_harris_wgan_benchmark_cs256_4gpus_layernorm_higherlr_**:
- `RMSE=1.1949 K`
-  `grad_ampl_ratio=0.9115` 

For an individual ensemble member, we get:
**_harris_wgan_benchmark4_t2m_rw1000_filter256_**: 
- `RMSE=1.3031 K`
- `grad_ampl_ratio=0.9538`

**_harris_wgan_benchmark_cs256_4gpus_layernorm_higherlr_**:
- `RMSE=1.4899 K`
-  `grad_ampl_ratio=1.0092` 

Thus, the Harris WGAN with batch normalization (checkpointed after 14 epochs) performs better in terms of per-ensemble member RMSE, but is inferior to the Harris WGAN with layer normalization (checkpointed after 28 epochs) in terms of spatial variability (per-ensemble member and ensemble mean). It's noteworthy that the inter-ensemble-spread with the Harris WGAN using layer normalization seems to be much higher, while ensemble-averaging reduces the RMSE much stronger compared to the Harris WGAN with batch normalization.
Furthermore, it is noted that `harris_wgan_benchmark_cs256_4gpus_layernorm_higherlr` has dignificantly better scores in the RALSD-evaluation.

Both models are however less performant than the 'old model configuration' (RMSE=1.12 K for the ensemble mean and RMSE=1.23 K for an individual ensemble member) which lacked Convolutional layers in the residual block and a complete residual block in the critic (cf. #46). 

We keep both trained model checkpoints for now, since we aim to keep the model configuration as consistent as possible for all experiments, and it's unclear yet how both model configurations behave for the different downscaling tasks (expecting potential training stability issues).

**Comment by mlangguth89** [comment]
First results from the latest model set-up (incl. layer normalization in the last non-linear Dense layer of the critic model, cf. #49):
After 14 epochs, the results for the 2m temperature seem to be improved alrady a bit compared to the two model checkpoints mentioned above.
For the ensemble mean, we get:
_**harris_wgan_benchmark_cs256_4gpus_layernorm_added_norm**_:
- `rmse=1.1725 K`
- `grad_ampl_ratio=0.9154`

For an individual ensemble member, we get:
- `rmse = 1.3059 K`
- `grad_ampl_ratio=1.0337`

Resuming training after 15 epochs improves slightly the RMSE (at the expense of the spatial variability):
For the ensemble mean, we obtain after 28 epochs:
- `rmse = 1.1660 K`
- `grad_ampl_ratio=0.8979`
For an individual ensemble member, we get:
- `rmse = 1.3179 K `
- `grad_ampl_ratio= 1.0050`

**Comment by ankitpatnala** [comment]
I was tracking some training loss values for both old and new dataset during training

```
old_data: 
    run_id: 11308266
   epoch_0: train_loss_step=0.142   train_loss_epoch=0.226
   epoch_21: train_loss_step=0.0973 train_loss_epoch=0.110 val_loss=0.137

new_data:
   run_id: 11307663
   epoch_0: train_loss_step=0.513   train_loss_epoch=3.030
   epoch_21: train_loss_step=0.295 train_loss_epoch=0.534 val_loss=0.653

old_data - **sp_in**: 
    run_id: 11308323
   epoch_0: train_loss_step=0.145   train_loss_epoch=0.215
   epoch_21: train_loss_step=0.105 train_loss_epoch=0.113 val_loss=0.144

new_data - **msl**:
   run_id: 11308294
   epoch_0: train_loss_step=0.901   train_loss_epoch=4.090
   epoch_21: train_loss_step=0.239 train_loss_epoch=0.331 val_loss=0.318
```

the only difference between old data and new data is old data has predictor variable `sp_in` and new_data had variable `msl_in` instead. Despite this, the loss values are very different. Even I tried removing these variable from respective dataset and the loss function still looked different. 

**In all our experiments, we deleted z_branch and output_channel was 1 i.e. t2m_tar**

**Comment by epavel1** [comment]
The results for the sha wgan model and the deepru models are depicted in the image below.
![Image](https://github.com/user-attachments/assets/afc1af64-39f9-47b3-9e46-011ad48580ad)

**Comment by mlangguth89** [comment]
For the Harris WGAN, epoch 28 from the trained `harris_wgan_benchmark_cs256_4gpus_layernorm_added_norm` will serve as the baseline model. Some basic evaluation is documented in [this hedgedoc](https://gitlab.jsc.fz-juelich.de/hedgedoc/8gygswPlSZaW_qelwX9z_Q#).

The respective model has been copied to `/p/project/hclimrep/downscale_bench/trained_models/t2m/`. The results are shared here `/p/project/hclimrep/downscale_bench/results/t2m/`

**Comment by ankitpatnala** [comment]
> For the Harris WGAN, epoch 28 from the trained `harris_wgan_benchmark_cs256_4gpus_layernorm_added_norm` will serve as the baseline model. Some basic evaluation is documented in [this hedgedoc](https://gitlab.jsc.fz-juelich.de/hedgedoc/8gygswPlSZaW_qelwX9z_Q#).
> 
> The respective model has been copied to `/p/project/hclimrep/downscale_bench/trained_models/t2m/`. The results are shared here `/p/project/hclimrep/downscale_bench/results/t2m/`

The hedgedoc is private

**Comment by epavel1** [comment]
## T2M Dataset – Benchmark Model Results

The models for the **T2M dataset** have been trained.

### Selected Benchmark Models

| Model      | Selected Runs |
|------------|----------------|
| Sha WGAN   | 3              |
| DeepRU     | 1              |
| Sha UNet   | 3              |

### Performance Metrics

| Model      | Grad Amplitude | RMSE       |
|------------|----------------|------------|
| DeepRU     | 0.8672         | 1.0382 K   |
| Sha WGAN   | 0.9500         | 1.0537 K   |
| Sha UNet   | 0.8667         | 1.0306 K   |
| SwinIR   | 0.8648        | 1.0519 K   |

![Image](https://github.com/user-attachments/assets/998593f3-5ea5-44f1-9dd4-96318c122e9b)

**Comment by mlangguth89** [comment]
@seblehner 
Shall we close this already?
Or shall we better await final models from SAMOS?

**Comment by seblehner** [comment]
Closing similar to https://github.com/mlangguth89/downscaling_benchmark/issues/49#issuecomment-3112352435.
